File: Answer/3418.机器人可以获得的最大金币数.cs
/*
 * @lc app=leetcode.cn id=3418 lang=csharp
 *
 * [3418] 机器人可以获得的最大金币数
 */

// @lc code=start
public class Solution
{
    public int MaximumAmount(int[][] coins)
    {
        var m = coins.Length;
        var n = coins[0].Length;
        return n <= m ? Solve(m, n, false) : Solve(n, m, true);

        int Solve(int rows, int cols, bool transposed)
        {
            const int NegInf = int.MinValue / 4;
            var dp0 = new int[cols];
            var dp1 = new int[cols];
            var dp2 = new int[cols];
            Array.Fill(dp0, NegInf);
            Array.Fill(dp1, NegInf);
            Array.Fill(dp2, NegInf);

            for (var r = 0; r < rows; r++)
            {
                var left0 = NegInf;
                var left1 = NegInf;
                var left2 = NegInf;
                for (var c = 0; c < cols; c++)
                {
                    var up0 = dp0[c];
                    var up1 = dp1[c];
                    var up2 = dp2[c];

                    int best0;
                    int best1;
                    int best2;
                    if (r == 0 && c == 0)
                    {
                        best0 = 0;
                        best1 = 0;
                        best2 = 0;
                    }
                    else
                    {
                        best0 = Math.Max(up0, left0);
                        best1 = Math.Max(up1, left1);
                        best2 = Math.Max(up2, left2);
                    }

                    var coin = transposed ? coins[c][r] : coins[r][c];
                    var cur0 = best0 + coin;
                    var cur1 = Math.Max(best1 + coin, best0);
                    var cur2 = Math.Max(best2 + coin, best1);

                    dp0[c] = cur0;
                    dp1[c] = cur1;
                    dp2[c] = cur2;

                    left0 = cur0;
                    left1 = cur1;
                    left2 = cur2;
                }
            }

            var last = cols - 1;
            return Math.Max(dp0[last], Math.Max(dp1[last], dp2[last]));
        }
    }
}
// @lc code=end

File: Answer/3418.机器人可以获得的最大金币数.java
/*
 * @lc app=leetcode.cn id=3418 lang=java
 *
 * [3418] 机器人可以获得的最大金币数
 */

// @lc code=start
class Solution {

    public int maximumAmount(int[][] coins) {
        int m = coins.length;
        int n = coins[0].length;
        if (n <= m) {
            return solve(coins, m, n, false);
        }
        return solve(coins, n, m, true);
    }

    private int solve(int[][] coins, int rows, int cols, boolean transposed) {
        final int negInf = Integer.MIN_VALUE / 4;
        int[] dp0 = new int[cols];
        int[] dp1 = new int[cols];
        int[] dp2 = new int[cols];
        for (int c = 0; c < cols; c++) {
            dp0[c] = negInf;
            dp1[c] = negInf;
            dp2[c] = negInf;
        }

        for (int r = 0; r < rows; r++) {
            int left0 = negInf;
            int left1 = negInf;
            int left2 = negInf;
            for (int c = 0; c < cols; c++) {
                int up0 = dp0[c];
                int up1 = dp1[c];
                int up2 = dp2[c];

                int best0;
                int best1;
                int best2;
                if (r == 0 && c == 0) {
                    best0 = 0;
                    best1 = 0;
                    best2 = 0;
                } else {
                    best0 = Math.max(up0, left0);
                    best1 = Math.max(up1, left1);
                    best2 = Math.max(up2, left2);
                }

                int coin = transposed ? coins[c][r] : coins[r][c];
                int cur0 = best0 + coin;
                int cur1 = Math.max(best1 + coin, best0);
                int cur2 = Math.max(best2 + coin, best1);

                dp0[c] = cur0;
                dp1[c] = cur1;
                dp2[c] = cur2;

                left0 = cur0;
                left1 = cur1;
                left2 = cur2;
            }
        }

        int last = cols - 1;
        return Math.max(dp0[last], Math.max(dp1[last], dp2[last]));
    }
}
// @lc code=end

